### ISSUE TITLE
Cannot build Syndesis with Operator

### ISSUE THREAD

**Issue by oscerd** [syndesisio/syndesis#3260]
## This is a...

<pre><code>
[ ] Feature request
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Documentation issue or request
</code></pre>

## The problem
`tools/bin/syndesis: line 95: cd: tools/bin: No such file or directory
Creating syndesis-operator with syndesis/godev:1.10
/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/util/operator_funcs: line 63: pushd: //install/operator: No such file or directory
`

## Expected behavior
build operator and install

**Comment by oscerd** [comment]
My path is jboss-fuse/syndesis/...

and this is the same even if I run the command from inside tools/bin/

**Comment by rhuss** [comment]
@oscerd can you run please with `--verbose` and send me the output around this error ?

**Comment by oscerd** [comment]
here is the output by the way:

`[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 11:45 min
[INFO] Finished at: 2018-08-01T16:33:50+02:00
[INFO] Final Memory: 461M/998M
[INFO] ------------------------------------------------------------------------
+++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:49): build::run(): should_build operator ''
+++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:214): should_build(): local module=operator
+++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:215): should_build(): local modules=
+++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:216): should_build(): '[' -z '' ']'
+++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:218): should_build(): echo true
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:49): build::run(): true
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:50): build::run(): operator
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:56): operator(): echo ==============================================================================
==============================================================================
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:57): operator(): echo 'Building syndesis-operator'
Building syndesis-operator
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:58): operator(): echo ==============================================================================
==============================================================================
+++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:60): operator(): appdir .
+++(./syndesis:100): appdir(): local subdir=.
++++(./syndesis:101): appdir(): basedir
++++(./syndesis:78): basedir(): local script=./syndesis
++++(./syndesis:81): basedir(): '[' -L ./syndesis ']'
+++++(./syndesis:94): basedir(): dirname ./syndesis
++++(./syndesis:94): basedir(): local dir=.
+++++(./syndesis:95): basedir(): cd .
+++++(./syndesis:95): basedir(): pwd
++++(./syndesis:95): basedir(): local full_dir=/home/oscerd/workspace/jboss-fuse/syndesis/app
++++(./syndesis:96): basedir(): echo /home/oscerd/workspace/jboss-fuse/syndesis/app
+++(./syndesis:101): appdir(): local dir=/home/oscerd/workspace/jboss-fuse/syndesis/app
++++(./syndesis:102): appdir(): cd /home/oscerd/workspace/jboss-fuse/syndesis/app/../../.
++++(./syndesis:102): appdir(): pwd
+++(./syndesis:102): appdir(): local full_dir=/home/oscerd/workspace/jboss-fuse
+++(./syndesis:103): appdir(): echo /home/oscerd/workspace/jboss-fuse
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:60): operator(): local topdir=/home/oscerd/workspace/jboss-fuse
+++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:61): operator(): hasflag -e --ensure
+++(./syndesis:108): hasflag(): filters='-e --ensure'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' build = -e ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' build = --ensure ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --openshift = -e ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --openshift = --ensure ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --clean = -e ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --clean = --ensure ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --verbose = -e ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --verbose = --ensure ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --skip-tests = -e ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --skip-tests = --ensure ']'
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:61): operator(): '[' '' ']'
+++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:64): operator(): hasflag -l --local
+++(./syndesis:108): hasflag(): filters='-l --local'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' build = -l ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' build = --local ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --openshift = -l ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --openshift = --local ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --clean = -l ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --clean = --local ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --verbose = -l ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --verbose = --local ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --skip-tests = -l ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --skip-tests = --local ']'
+++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:64): operator(): hasflag -i --image --images
+++(./syndesis:108): hasflag(): filters='-i --image --images'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' build = -i ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' build = --image ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' build = --images ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --openshift = -i ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --openshift = --image ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --openshift = --images ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --clean = -i ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --clean = --image ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --clean = --images ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --verbose = -i ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --verbose = --image ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --verbose = --images ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --skip-tests = -i ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --skip-tests = --image ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --skip-tests = --images ']'
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:64): operator(): build_operator /home/oscerd/workspace/jboss-fuse '' ''
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/util/operator_funcs:44): build_opelocal top_dir=/home/oscerd/workspace/jboss-fuse
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/util/operator_funcs:45): build_opelocal do_local=
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/util/operator_funcs:46): build_opelocal do_linux=
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/util/operator_funcs:48): build_ope'[' '' ']'
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/util/operator_funcs:62): build_opeecho 'Creating syndesis-operator with syndesis/godev:1.10'
Creating syndesis-operator with syndesis/godev:1.10
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/util/operator_funcs:63): build_opepushd /home/oscerd/workspace/jboss-fuse/install/operator
/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/util/operator_funcs: line 63: pushd: /home/oscerd/workspace/jboss-fuse/install/operator: No such file or directory
`

**Comment by rhuss** [comment]
@oscerd could you please retry as #3262 is merged ?

**Comment by oscerd** [comment]
I'm trying again now.

**Comment by oscerd** [comment]
In my env it fails anyway

```+++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:51): build::run(): should_build operator ''
+++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:216): should_build(): local module=operator
+++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:217): should_build(): local modules=
+++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:218): should_build(): '[' -z '' ']'
+++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:220): should_build(): echo true
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:51): build::run(): true
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:52): build::run(): operator /home/oscerd/workspace/jboss-fuse/syndesis
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:58): operator(): local top_dir=/home/oscerd/workspace/jboss-fuse/syndesis
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:59): operator(): echo ==============================================================================
==============================================================================
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:60): operator(): echo 'Building syndesis-operator'
Building syndesis-operator
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:61): operator(): echo ==============================================================================
==============================================================================
+++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:63): operator(): hasflag -e --ensure
+++(./syndesis:108): hasflag(): filters='-e --ensure'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' build = -e ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' build = --ensure ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --openshift = -e ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --openshift = --ensure ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --clean = -e ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --clean = --ensure ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --verbose = -e ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --verbose = --ensure ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --skip-tests = -e ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --skip-tests = --ensure ']'
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:63): operator(): '[' '' ']'
+++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:66): operator(): hasflag -l --local
+++(./syndesis:108): hasflag(): filters='-l --local'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' build = -l ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' build = --local ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --openshift = -l ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --openshift = --local ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --clean = -l ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --clean = --local ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --verbose = -l ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --verbose = --local ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --skip-tests = -l ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --skip-tests = --local ']'
+++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:66): operator(): hasflag -i --image --images
+++(./syndesis:108): hasflag(): filters='-i --image --images'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' build = -i ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' build = --image ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' build = --images ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --openshift = -i ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --openshift = --image ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --openshift = --images ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --clean = -i ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --clean = --image ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --clean = --images ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --verbose = -i ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --verbose = --image ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --verbose = --images ']'
+++(./syndesis:109): hasflag(): for var in '"${ARGS[@]}"'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --skip-tests = -i ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --skip-tests = --image ']'
+++(./syndesis:110): hasflag(): for filter in '$filters'
+++(./syndesis:111): hasflag(): '[' --skip-tests = --images ']'
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/build:66): operator(): build_operator /home/oscerd/workspace/jboss-fuse/syndesis '' ''
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/util/operator_funcs:44): build_opelocal top_dir=/home/oscerd/workspace/jboss-fuse/syndesis
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/util/operator_funcs:45): build_opelocal do_local=
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/util/operator_funcs:46): build_opelocal do_linux=
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/util/operator_funcs:48): build_ope'[' '' ']'
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/util/operator_funcs:62): build_opeecho 'Creating syndesis-operator with syndesis/godev:1.10'
Creating syndesis-operator with syndesis/godev:1.10
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/util/operator_funcs:63): build_opepushd /home/oscerd/workspace/jboss-fuse/syndesis/install/operator
++(/home/oscerd/workspace/jboss-fuse/syndesis/tools/bin/commands/util/operator_funcs:64): build_opedocker run -w /gopath/src/github.com/syndesisio/syndesis/install/operator -v /home/oscerd/workspace/jboss-fuse/syndesis:/gopath/src/github.com/syndesisio/syndesis:z -e CGO_ENABLED=0 syndesis/godev:1.10 go build ./cmd/syndesis-operator
container_linux.go:247: starting container process caused "chdir to cwd (\"/gopath/src/github.com/syndesisio/syndesis/install/operator\") set in config.json failed: no such file or directory"
/usr/bin/docker-current: Error response from daemon: invalid header field value "oci runtime error: container_linux.go:247: starting container process caused \"chdir to cwd (\\\"/gopath/src/github.com/syndesisio/syndesis/install/operator\\\") set in config.json failed: no such file or directory\"\n".
ERRO[0000] error getting events from daemon: net/http: request canceled ```

**Comment by oscerd** [comment]
Probably I have to move my syndesis repo under my gopath

**Comment by oscerd** [comment]
No, copying the stuff there won't work. The gopath is used in docker during build so don't know what is the problem.

**Comment by oscerd** [comment]
I'm reopening this just because it's related.

**Comment by rhuss** [comment]
No, copying to GOPATH would be only required if you call it locally with the option `--local`. 

It looks like that a volume moutn doesn't work for you. How do you access you Docker daemone ? Is it running on your machine (linux, localhost) or do you use the Docker daemon of minishift which runs in a VM ?

Actually could you try this command locally:

```
        docker run -w /gopath/src/github.com/syndesisio/syndesis/install/operator \
                   -v $(pwd)/dep-cache:/gopath/pkg/dep:z \
                   -v $top_dir:/gopath/src/github.com/syndesisio/syndesis:z \
                    \
                   dep ensure -vendor-only -v
```

with your current dir in `tools/operator` and `$top_dir` set to `/home/oscerd/workspace/jboss-fuse/syndesis/`

**Comment by oscerd** [comment]
It is running on minishift.

I do eval $(minishift docker-env)

**Comment by oscerd** [comment]
What's the procedure on minishift for a local build? I always used remote mod.

**Comment by rhuss** [comment]
We do the binary build for the Docker image of the operator (image syndesis-operator). But compiling it would be a source S2I build (we have no binary yet), and moving deps around (with 400 MBs) will be an awful experience I'm afraid.

This issue is all about compiling and running dep before we can put the operator into an image.

**Comment by rhuss** [comment]
@oscerd it looks like this for me if the $GOPATH is set properly:

```
mkdir -p $GOPATH/src/github.com/syndesiso
mv /home/oscerd/workspace/jboss-fuse/syndesis $GOPATH/src/github.com/syndesiso/
ln -s $GOPATH/src/github.com/syndesiso/syndesis  /home/oscerd/workspace/jboss-fuse/
syndesis build -m operator --local --ensure # or -l -e
```

**Comment by rhuss** [comment]
Ah, I forgot: For `--ensure` you need to be somewhere in `$GOPATH/src/github.com/syndesiso/syndesis` (afair being in the symlinked directory doesnt work as `dep ensure` moves up the directory path, not respecting symlinks).

But you will find out whether this is necessary (and let me know so that I can add it to the docs)

**Comment by rhuss** [comment]
@nicolaferraro for building in a remote Docker daemon I thought about using something loke

```
docker volume create dep-cache  # but only if not already exists
docker run -v dep-cache:/gopath/pkd/dep .....
```

Then, as long as the volume is not pruned, its holds the dependencies. Same for the vendor dir. The only tricky part is, how to get binary out of the container when compilation is done.

Multi-stage builds also don't work as we need the binary for an S2I build probably.

**Comment by oscerd** [comment]
So I just need to run the syndesis with -l and -e with local docker

**Comment by oscerd** [comment]
I never tried this approach I always use a remote daemon

**Comment by rhuss** [comment]
@oscerd no, -l is for running real locally without any Docker Daemon. You would need to have `go` and `dep` installed so that it can be called from the script.

If you have a local Docker daemon (i.e. running on your host, not in a VM or in VM which can do bind mounts to your local host), the you omit the `-l`. So, that would you did would work on a local Docker daemon.

When you run on Mac's Docker this is like a local Docker daemon. When you run on Linux, you could easily install a local docker daemon, too.

But let me check whether it can be possible for Minishift to do bind mounts to the host outside the VM.

**Comment by rhuss** [comment]
@oscerd @nicolaferraro I think I found a possible solution, specifically for compiling the binary with `minishift` --> https://docs.openshift.org/latest/minishift/using/host-folders.html

So let me introduce a flag `syndesis build --minishift` (or an autodetection on the `DOCKER_HOST` variable) and then use hostfolder with minishift. I think that is all that I can do for supporting to build on a remote Docker daemon (so it will only work with minishift when being a remote daemon).

**Comment by oscerd** [comment]
Ok thanks.

**Comment by rhuss** [comment]
@oscerd actually I tried hard, but couldn't succeed: First, minishift's host folders are vias sshfs so possibly dead slow when it comes to compiling over it. Second, I fail to volume mount such a mounted folder within a Docker contaienr because of SELinux restrictions. 

I could dig further, but decided now its better to document first the behaviour tommorrow and then maybe continue here.

Also, I will introduce an options like --all-modules which will build everything, otherwise only the java build is performed. Actually, you only have to redeploy the operator if you are working on the operator (otherwise its pulled from Docker Hub anyway). And someone working on the operator possible has a local go environment anyway (because that is the most performant one).

For now I just have to say sorry for setup without a local Docker daemon and without a go developer setup ;-(

sorry ...

**Comment by oscerd** [comment]
Ok, thanks, for the moment I'll build with the last updated image and then rebuild locally only what I need.

**Comment by rhuss** [comment]
I'm just evaluating use rsync with minishift to get the source into the minishift vm (and the binary out again). Looks promising ...

**Comment by rhuss** [comment]
@heiko-braun I think its ok to lower the priority as this build issue happens only when using the Minishift Docker daemon to build the operator image. There are two other options (local go, local Docker daemon) which can be used. The release and CI are not affected as they are using local Docker daemons.

While I'm trying to fix this for Minishift, too, the solution will be for sure slower then the other two (with the local go build being the fastest). 

The operator needs to be build only by those who are developing the operator, otherwise the latest Docker image of the operator should be picked up from Docker Hub anyway.

tl;dr - I don't consider this to be a blocker as there other, more recommended alternatives to build the operator.

**Comment by lburgazzoli** [comment]
@rhuss so maybe remove the operator from --all-images and add a custom flag to build it

**Comment by rhuss** [comment]
yes, that's planned with my next PR. Opertor building needs to be switched on on demand.

**Comment by lburgazzoli** [comment]
maybe a custom command also syndesis operator --build|--deploy (etc) so it also less confusing for standard build and can also take into accounto additional operators we may add in teh future

**Comment by rhuss** [comment]
Yeah, of course a good idea as that was my first though that I throw into #syndesis but was rejected ;-)

But that's no problem at all to have both as the operator specific parts have been already extracted into an extra script file which can be included by both.

Love the revival of one of my favourite the Perl mantras [TIMTOWTDI](http://wiki.c2.com/?ThereIsMoreThanOneWayToDoIt)

**Comment by zregvart** [comment]
I would prefer to stick with `syndesis build` i.e. `syndesis build -m operator`, if I need to rebuild everything I don't want to run `syndesis build -c` and then `syndesis operator --build` because I'll forget the latter.

**Comment by lburgazzoli** [comment]
We can of course keep it but I found it confusing as we have flags that are unrelated i.e --local and --ensure do not make any sense for other builds than the operator but I can get used to of course

**Comment by rhuss** [comment]
I fixed the build issue with the Minishift docker daemon with #3272 and made some timing about the build for the three ways for `syndesis build -m operator --ensure`

| Context | Local | Local Docker Daemon (OSX) | Minishift Docker Daemon  |
| ------- | ----- | ------------------------- | ------------------------ |
| Cold (no build cache) | 0:40 | 10:09 | 1:36 |
| Hot (with dep cache) | 0:08 | 2:47 | 0:13 |

Not sure why OSX Docker is so slow, I probably have to restart that. Shouldn't be that slow, but Minishift build is decently fast.

**Comment by rhuss** [comment]
> We can of course keep it but I found it confusing as we have flags that are unrelated i.e --local and --ensure do not make any sense for other builds than the operator but I can get used to of course

The same is btw true for `--flash` etc. for the operator build.

My sugesstion:

* **syndesis build**: 

```
-b  --backend                 Build only backend modules (core, extension, integration, connectors, server, meta)
    --all-images              Build all modules with images: ui, server, meta, s2i, operator
    --java-images             Build only Java modules with Docker images (ui, server, meta, s2i) and create images
-m  --module <m1>,<m2>, ..    Build modules
                              Modules: ui, server, connector, s2i, meta, integration, extension, common, operator
-d  --dependencies            Build also all project the specified module depends on
    --skip-tests              Skip unit and system test execution
    --skip-checks             Disable all checks
-f  --flash                   Skip checks and tests execution (fastest mode)
-i  --image                   Build Docker via s2i images, too
    --docker                  Use a plain Docker build for creating images. Used by CI for pushing to Docker Hub
-p  --project <project>       Specifies the project to create images in when using '--openshift'
-e  --ensure                  If building the operator , run 'dep ensure' before building. Otherwise ignored.
-l  --local                   If building the operator, use a locally installed go
                              Otherwise run the Go build from a container (either local or Minishift's Docker daemon)
-c  --clean                   Run clean builds (mvn clean)
    --resume <m1>             Resume build from maven module (mvn -rf m1)
    --batch-mode              Run mvn in batch mode
    --camel-snapshot          Run a build with a specific Camel snapshot. You'll need to set an environment variable CAMEL_SNAPSHOT_VERSION with the SNAPSHOT version you want to use.
```

* **syndesis operator**

```
    --build                   Build operator
-e  --ensure                  Run 'dep ensure' before building
-l  --local                   Use a locally installed go for building
                              Otherwise run the Go build from a container (either local or Minishift's Docker daemon)
-i  --image                   Create Docker image with S2I
-d  --docker                  Create Docker image via a Docker Daemon
    --install                 Install the operator in the currently connected project and
                              creates a "Syndesis" resource if not present
    --setup-crd               Setup CRD for the connected cluster. Requires cluster-admin
    --grant <user>            Grant extra permission required by <user> for installing
                              the operstor. Requires cluster-admin. Used with --install
```

@zregvart @lburgazzoli wdyt ? This should fit both use cases (except that '--all-images' becoms now '--java-images'

The only question is what to do when no modules are selected: Build only Java modules or also the operator ? (e.g. for `syndesis build -f -c`). I would say without operator.

**Comment by rhuss** [comment]
cc @syndesisio/all If this is good for everyone I would implement and document it that way.

**Comment by lburgazzoli** [comment]
I'd say:

--app-images
--infra-images

As UI is not really a java thing :)

Without selector I think it should build everything

**Comment by rhuss** [comment]
sounds legit, thanks !

**Comment by zregvart** [comment]
I don't like that we have another _build_ command now one for backend/ui and another for the operator. We should have a single _build_ command for all.

So, cant `dep ensure` be run on every build? Can't we detect if there is `go` installed locally and then build locally else build in Docker?

**Comment by rhuss** [comment]
@zregvart you can still use 'syndesis build -m operator'. I uses the same functions as 'sydesis operator --build'

**Comment by rhuss** [comment]
@zregvart wrt to automatic build detection: Having a local go installation doesn't mean that you are properly setup for building locally. There is so many things that go wrong (wrong go version). Also for the CI I want it to *always* run with a container for reproducability as this is the way how the release works.

Running "dep ensure" costs considerable time, not really suitable for development. But I will add an implicit dep ensure when an error occurs because of missing deps.

**Comment by zregvart** [comment]
@rhuss I'm then missing the point, why have `syndesis operator` command if it's functionality is present in `syndesis build` and `syndesis [minishift --install|install]`?

**Comment by rhuss** [comment]
yes, its just approaching it differently, for someone working mostly or exclusively on operators. All other can ignore 'syndesis operator'.

**Comment by rhuss** [comment]
ok, compromise: Let me fix 'syndesis build' and 'syndesis install' (for '--admin-setup') and then we can still discuss whether we want to have 'syndesis operator'.

Deal?
